doc: move Package Porting Tutorial to Guides and extend it (#22254)
RIOT-OS/RIOT: doc/guides/advanced_tutorials/porting_packages.mdx

Your package folder can also contain additional source or header files that
are required for the package to work.
Due to the heterogenous nature of external packages, this is highly specific
to the package in question. It is best to look at other packages and their
use of additional source or header files to understand how to use this
feature.

To include the additional headers, you have to create a `Makefile.include` as
discussed above.

### Using a Local Copy for Development

The build system will always pull a fresh copy of the package from the git
repository or `http` source linked in your package's `Makefile`.

This can make the development challenging, as every local change will be
overridden by a new compile run, leading to data loss and potentially the loss
of many hours of work.

To avoid having to set up a separate development repository, you can specify a
local folder using the `PKG_SOURCE_LOCAL` variable.

You can either specify it temporarily in your package's `Makefile` or add it
to the compiler call when building your test application, for example:

```shell
PKG_SOURCE_LOCAL=~/customized_package_folder make -C tests/pkg/new_pkg_test

> ```

Author: I should remove or reword this because `pkg/pkg.mk` will use `PKG_SOURCE_LOCAL` for *all* packages if multiple packages are used.

So also for `CMSIS` for example etc.

Author: Today I learned: you can also specify the variable for individual packages 🤯 

See #11533, I'll extend the documentation and add a warning to `pkg/pkg.mk`.

Author: Added.

person available to help with potential issues that surface later on.

## Structure of a package

If you'd like to add a package to RIOT, you need to add a directory with the
name of your package to the `pkg/` directory.
The directory structure should look like this, and the files marked with an
asterisk are mandatory.

<FileTree>

- \<pkgname\>
    - patches
        - 0001-your-fancy-changes.patch
    - doc.md*
    - Makefile*
    - Makefile.dep
    - Makefile.include

</FileTree>

### doc.md (required)

To help people use your package, it is required to add a `doc.md` Markdown file
to your package folder. Many packages have configuration options, other
dependencies or limitations that might make it non-trivial to use for someone
who is not yet familiar with the package.

### Makefile (required)

The `Makefile` describes how to fetch the upstream application,
apply the patch(es) and how to build the package as a RIOT module.
Two rough templates for acquiring a package via `git` or downloading it via
`http` are provided in `pkg/Makefile.git` and `pkg/Makefile.http`.

Depending on the code quality of the package, you might have to disable some

> compiler warnings as well by extending the `CFLAGS` variable, for example:

Reviewer: ```suggestion
Depending on the code quality of the package, you might have to disable some
compiler warnings when building the package by extending the `CFLAGS` variable, for example:
```

```makefile
CFLAGS += -Wno-sign-compare
CFLAGS += -Wno-strict-aliasing
```

### Makefile.include

> If your port or the package provide header files, you need to extend the

Reviewer: ```suggestion
If your port or the package provide header files which you want to include from RIOT or application code, you need to extend the
```

`INCLUDES` variable in the package's `Makefile.include`. This file is optional.

```makefile
INCLUDES += -I$(RIOTPKG)/<pkg_name>/include # headers provided by the port
INCLUDES += -I$(PKGDIRBASE)/<pkg_name> # headers provided by the package
```

To avoid `clang` checking the documentation and integrety of the external
headers, you can use `-Isystem$(PKGDIRBASE)/<pkg_name>` instead. This will

> treat the headers as System Headers and not headers that are part of your

Reviewer: ```suggestion
treat the headers as system headers and not headers that are part of your
```

files etc.

:::note
Remember to remove the `PKG_SOURCE_LOCAL` variable from your package's
`Makefile` before submitting a Pull Request!
:::

## Creating a Patch with git

Assuming your upstream library resides in a git repository, you can create the
patch files as follows:
 - checkout the targeted version of the upstream application
 - create a new branch (e.g. `riot-port` by executing `git checkout -b riot-port`)
 - conduct necessary changes (e.g. edit, add, or remove some files)
 - commit your changes using `git commit`. Remember to add a detailed
   description in your commit, as this will be added to the patch as well.
 - create the patch files using `git format-patch -n riot-port`
 - move the resulting patch files to the `patches/` directory of your package
 - see [patches/](#patches) for more information about how to name your
   patch files

Try to keep the patches of reasonable size to help with the review process.
Patches often have to be updated when changes in the upstream repository were
made and having specific patches can make the update process a lot easier.

If you have discovered any bugs or changes that the upstream repository would

> also benefit from, you should of course publish a Pull Request and keep the

Reviewer: ```suggestion
also benefit from, you should of course create a Pull Request at the upstream repository and keep the
```

patch singled out. Once your Pull Request was merged, you can remove the
individual patch file.

## Packages outside of RIOTPKG

It can be beneficial to create packages outside of the RIOT tree. For example
if one is working on new packages that aren't ready to be committed to
upstream or if an application needs its own unique packages. For this, one
can use the `EXTERNAL_PKG_DIRS` make variable. It works similar to the way
[external modules](https://guide.riot-os.org/advanced_tutorials/creating_modules/#modules-outside-of-riotbase)
are handled. In your application's `Makefile`, in addition to adding the package
name to `USEPKG` as shown above, add the path to a folder that contains your
external packages:

```makefile
EXTERNAL_PKG_DIRS += <PATH_TO_FOLDER_CONTAINING_PACKAGES>
```

The path is allowed to be relative to the application's directory.

:::note
The name of an external package must be unique (both in regard to
other external packages, as well to native RIOT packages). Additionally, the
directory containing the packages must match the package name, e.g. package

> `foo`must be located in `<PATH_IN_EXTERNAL_PKG_DIRS>/foo`.

Reviewer: ```suggestion
`foo` must be located in `<PATH_IN_EXTERNAL_PKG_DIRS>/foo`.
```

:::

An example can be found in
[`tests/build_system/external_pkg_dirs`](https://github.com/RIOT-OS/RIOT/tree/master/tests/build_system/external_pkg_dirs)

## How the Build System handles Packages

> ***TODO:*** This section should have more information about how `pkg/pkg.mk` works

Reviewer: Is this a TODO to be solved before merging?

Author: Not really, I wanted to put a break somewhere to avoid it becoming an endless PR. I'd rather open another PR when I want to put more work into it (or someone else for that matter) :D

internally and the steps of preparing a package.

When the package can be built out-of-source, the source code of external
packages is fetched in a global location in:

```makefile
$(PKGDIRBASE)/$(PKG_NAME)
```

When out-of-source build is not possible (for example because the package
build system doesn't allow it), the source code is fetched in the build
directory of the application under:

```makefile

> $(BINDIR)/pkg/$(PKG_NAME)

Reviewer: In case `$BUILD_DIR` is set (globally) for the RIOT compilation, it will instead end up in the (global) location `$(BUILD_DIR)/pkg/$(PKG_NAME)`

Reviewer: Or maybe this does not apply to micropython but only to the case above, not sure.

Author: Not sure, that was copied from the original guide.

Reviewer: Just tried, indeed for micropython it stays in this directory.

But you could update the line above to say that if `$(BUILD_DIR)` is set, it will end up below that path instead of `$(PKGDIRBASE)`.

Author: I rewrote the whole thing. For the first time I think I understood now how the `PKG_SOURCE_DIR` and `PKG_BUILD_DIR` stuff works 😅

and can be used to pull additional dependencies that are required by your
package. This file is optional.

### patches/

Patch files can be included in a `patches` directory in the package dir.
These are applied to the upstream package to make it build with RIOT.
The patch files should follow the general scheme `000x-descriptive-title.patch`,
some examples are given below.

```
0001-Initial-Changes.patch
0002-Fix-encoding-issues.patch
0003-Add-RIOT-porting-layer.patch
...
```

See [Creating a Patch with git](#creating-a-patch-with-git) for more
information about how to create a patch.

### Additional Sources and Headers

Your package folder can also contain additional source or header files that
are required for the package to work.
Due to the heterogenous nature of external packages, this is highly specific
to the package in question. It is best to look at other packages and their
use of additional source or header files to understand how to use this
feature.

To include the additional headers, you have to create a `Makefile.include` as
discussed above.

## Using git to fetch Packages

***TODO:*** This section should have more information about how `pkg/pkg.mk` works

> internally and the steps of preparing a package.

Reviewer: I think something has gone wrong here when moving the section o.O

Author: No, that was intentional. I removed the `How the build system handles blabla` section.

As `pkg/pkg.mk` is only relevant for `git`, this is about the right place.

example when the package is using CMake, the separate `PKG_BUILD_DIR`
directory should be used for the build process. The default path for
it is `$(BINDIR)/pkg-build/$(PKG_NAME)`.
If no separate build process is required, the `PKG_BUILD_DIR` can be left
unused.

Below is a typical file tree for out-of-source building.
`PKGDIRBASE` is typically set to `$(BUILD_DIR)/pkg` and `BUILD_DIR`
is set to `$(RIOTBASE)/build`.

<FileTree>
- $(RIOTBASE)
  - examples
    - \<application\>
      - bin = $(BINDIRBASE)
        - \<board\> = $(BINDIR)
          - pkg-build
            - \<pkgname\> = $(PKG_BUILD_DIR)
  - build = $(BUILD_DIR)
    - pkg = $(PKGDIRBASE)
      - \<pkgname\> = $(PKG_SOURCE_DIR)

</FileTree>

:::note
You can change `PKG_SOURCE_DIR` and `PKG_BUILD_DIR` in your package's
`Makefile`, but you should not change `PKGDIRBASE`, `BINDIR` or
`BUILD_DIR`, as this can quickly lead to unforeseen side effects.
:::

If your package can not be build out-of-source, you can set the
`PKG_BUILD_OUT_OF_SOURCE` variable to `0`. This will cause the buildsystem
to set `$(PKG_SOURCE_DIR) = $(PKG_BUILD_DIR) = $(BINDIR)/pkg/$(PKG_NAME)`.

> The `PKG_SOURCE_DIR` can be changed, while the `PKG_BUILD_DIR` is fixed.

Reviewer: So rather this way around, right?

```suggestion
to set `$(PKG_BUILD_DIR) = $(PKG_SOURCE_DIR) = $(BINDIR)/pkg/$(PKG_NAME)`.

The `PKG_SOURCE_DIR` can be changed, while the `PKG_BUILD_DIR` is fixed.
```

<FileTree>

- $(RIOTBASE)
  - examples
    - \<application\>
      - bin = $(BINDIRBASE)
        - \<board\> = $(BINDIR)
          - pkg
            - \<pkgname\> = $(PKG_SOURCE_DIR) = $(PKG_BUILD_DIR)

</FileTree>

Special precautions have been taken to avoid deleting source files, when the
`clean` target is called, so only intermediate build files are deleted.

In-source builds are the exception and only very few packages require it.

## How the Build System handles Packages

***TODO:*** This section should have more information about how `pkg/pkg.mk` works
internally and the steps of preparing a package.

> ## Using http to fetch Packages

Reviewer: Shouldn't this come directly below "using git to fetch packages"?

Author: Yeah probably.

If your package is not available in form of a repository, you can base your
package's `Makefile` on the `pkg/Makefile.http` example and fill out the
individual build steps that are required by your package.

The http based process is lacking many of the quality of life helpers that

> the git based process offers. For example, the automatic patching is currently

Reviewer: ```suggestion
The http-based process is lacking many of the quality-of-life features that
the git-based process offers. For example, the automatic patching is currently
```
